renderdoc/driver/d3d12/d3d12_shaderdebug.cpp

```diff
@@ -96,8 +96,176 @@ bool D3D12DebugAPIWrapper::CalculateMathIntrinsic(DXBCBytecode::OpcodeType opcod
                                                   const ShaderVariable &input,
                                                   ShaderVariable &output1, ShaderVariable &output2)
 {
-  RDCUNIMPLEMENTED("CalculateMathIntrinsic not yet implemented for D3D12");
-  return false;
+  D3D12MarkerRegion region(m_pDevice->GetQueue()->GetReal(), "CalculateMathIntrinsic");
+
+  rdcstr csProgram =
+      "RWStructuredBuffer<float4> outval : register(u0);\n"
+      "cbuffer srcOper : register(b0) { float4 inval; };\n"
+      "[numthreads(1, 1, 1)]\n"
+      "void main(){\n";
+
+  switch(opcode)
+  {
+    case DXBCBytecode::OPCODE_RCP: csProgram += "  outval[0] = rcp(inval);\n"; break;
+    case DXBCBytecode::OPCODE_RSQ: csProgram += "  outval[0] = rsqrt(inval);\n"; break;
+    case DXBCBytecode::OPCODE_EXP: csProgram += "  outval[0] = exp2(inval);\n"; break;
+    case DXBCBytecode::OPCODE_LOG: csProgram += "  outval[0] = log2(inval);\n"; break;
+    case DXBCBytecode::OPCODE_SINCOS:
+      csProgram += "  sincos(inval, outval[0], outval[1]);\n";
+      break;
+    default: RDCERR("Unexpected opcode %d passed to CalculateMathIntrinsic", opcode); return false;
+  }
+
+  csProgram += "}\n";
+
+  ID3DBlob *csBlob = NULL;
+  UINT flags = D3DCOMPILE_DEBUG | D3DCOMPILE_WARNINGS_ARE_ERRORS | D3DCOMPILE_DEBUG_NAME_FOR_SOURCE;
+  if(m_pDevice->GetShaderCache()->GetShaderBlob(csProgram.c_str(), "main", flags, "cs_5_0",
+                                                &csBlob) != "")
+  {
+    RDCERR("Failed to create shader to calculate math intrinsic");
+    return false;
+  }
+
+  // Create a fresh root signature, since the shader being debugged could be using the graphics pipe
+  D3D12RootSignature rootSig;
+
+  D3D12RootSignatureParameter constantParam;
+  constantParam.ParameterType = D3D12_ROOT_PARAMETER_TYPE_32BIT_CONSTANTS;
+  constantParam.ShaderVisibility = D3D12_SHADER_VISIBILITY_ALL;
+  constantParam.Constants.Num32BitValues = 4;    // Input is 4 floats
+  constantParam.Constants.ShaderRegister = 0;
+  constantParam.Constants.RegisterSpace = 0;
+  rootSig.Parameters.push_back(constantParam);
+
+  D3D12RootSignatureParameter uavParam;
+  uavParam.ParameterType = D3D12_ROOT_PARAMETER_TYPE_UAV;
+  uavParam.ShaderVisibility = D3D12_SHADER_VISIBILITY_ALL;
+  uavParam.Descriptor.ShaderRegister = 0;
+  uavParam.Descriptor.RegisterSpace = 0;
+  uavParam.Descriptor.Flags = D3D12_ROOT_DESCRIPTOR_FLAG_NONE;
+  rootSig.Parameters.push_back(uavParam);
+
+  ID3DBlob *root = m_pDevice->GetShaderCache()->MakeRootSig(rootSig);
```

Comment:
This would be a good candidate to create up front in `D3D12DebugManager` since this root signature doesn't vary by parameters.